libretroshare/src/services/p3gxschannels.cc

```diff
@@ -1036,6 +1042,108 @@ bool p3GxsChannels::getChannelsContent(
 	return getPostData(token, posts, comments);
 }
 
+bool p3GxsChannels::createChannel(RsGxsChannelGroup& channel)
+{
+	uint32_t token;
+	time_t beginCreation = time(nullptr);
+	if( !createGroup(token, channel)
+	        || waitToken(token) != RsTokenService::COMPLETE )
+		return false;
+	time_t endCreation = time(nullptr);
+
+	std::list<RsGroupMetaData> channels;
+	if(!getChannelsSummaries(channels)) return false;
+
+	/* This is ugly but after digging and doing many tries of doing it the right
+	 * way ending always into too big refactor chain reaction, I think this is
+	 * not that bad, moreover seems the last created group tend to end up near
+	 * the beginning of the list so it is fast founding it.
+	 * The shortcoming of this is that if groups with same data are created in
+	 * a burst (more then once in a second) is that the id of another similar
+	 * group can be returned, but this is a pointy case.
+	 * Order of conditions in the `if` matter for performances */
+	bool found = false;
+	for(const RsGroupMetaData& chan : channels)
+	{
+		if( IS_GROUP_ADMIN(chan.mSubscribeFlags)
```

Comment:
> Indeed, that is really ugly. Another way to go would be to add the creation token into RsGroupMetaData, or use some kind of preliminary identifier, returned by createGroup() that can further be used to retrieve the correct channel.

I have experimented various ways of doing that, the problem is that while we keep track of token and token status in RsGxsDataAccess also for channel creation like we do for data read request, RsGxsDataAccess cannot keep track of channel creation data and request because it is done into RsGenExchange so triyng to keep some data into RsGxsDataAccess about the channel a race condition between RsGxsDataAccess and RsGenExchange raise and the channel is never created (also RS need to be stopped by kill -9 in that case...)

The most elegant working solution I could think of after many experiments is to move the channel creation into RsGxsDataAccess and fully keep track of it there like we do with the other GxsRequests, but indeed attempting to to that caused a refactoring chain reaction, and I have no time for that right now and I think it should eventually done into it's own pull request

> use some kind of preliminary identifier

token is meant for that use but is not usable to retrieve data in this case because of what I explained before

Reply:
> Besides if the channel creation fails for some reason, a function calling getChannelSummary() indefinitly would never return.

Nope it would return false in that case quite fast, anyway in this context we would never call that function if channel creation failed because if `waitToken` return something different than `COMPLETE` `createChannel` return false before of getting there.

Also take in account that `waitToken` by default has a timeout of 500ms